File: proposed-projects/round-trip-experiment/reverse/20260909T032340Z_f01d5496/20260910T063907Z_1875e571/ported_implementation/src/utils/errors/build-error-position.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  while (lineIdx < lines.length && lines[lineIdx] && pos > lines[lineIdx].length - 1) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  const start = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
  for (let i = start; i <= lineIdx; i++) {
    // the reference implementation indexes past the end of the array here, which joins
    // as an empty string.
    linesToShow.push(i < lines.length ? lines[i] : '');
  }

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};

File: packages/gbnf-experiment/data/20260909T012246Z_fbe56704/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == "":
        return ["No input provided"]

    lines = src.split("\n")

    line_idx = 0
    while line_idx < len(lines) and lines[line_idx] and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: List[str] = []
    for i in range(max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)), line_idx + 1):
        # the reference implementation indexes past the end of the array here, which
        # renders as an empty line once joined.
        lines_to_show.append(lines[i] if i < len(lines) else "")

    return [
        *lines_to_show,
        " " * pos + "^",
    ]
